pkg/auth/auth.go

```diff
@@ -59,98 +99,152 @@ func NewConfig(usePodIdentity, useVMManagedIdentity bool, userAssignedIdentityID
 	config.UsePodIdentity = usePodIdentity
 	config.UseVMManagedIdentity = useVMManagedIdentity
 	config.UserAssignedIdentityID = userAssignedIdentityID
+	config.WorkloadIdentityClientID = workloadIdentityClientID
+	config.WorkloadIdentityToken = workloadIdentityToken
 
 	return config, nil
 }
 
-func (c Config) GetServicePrincipalToken(podName, podNamespace, resource, aadEndpoint, tenantID, nmiPort string) (*adal.ServicePrincipalToken, error) {
+// GetAuthorizer returns an Azure authorizer based on the provided azure identity
+func (c Config) GetAuthorizer(ctx context.Context, podName, podNamespace, resource, aadEndpoint, tenantID, nmiPort string) (autorest.Authorizer, error) {
+	if c.UsePodIdentity {
+		return getAuthorizerForPodIdentity(podName, podNamespace, resource, aadEndpoint, tenantID, nmiPort)
+	}
+	if c.UseVMManagedIdentity {
+		return getAuthorizerForManagedIdentity(resource, c.UserAssignedIdentityID)
+	}
+	if len(c.AADClientSecret) > 0 && len(c.AADClientID) > 0 {
+		return getAuthorizerForServicePrincipal(c.AADClientID, c.AADClientSecret, resource, aadEndpoint, tenantID)
+	}
+	if len(c.WorkloadIdentityToken) > 0 && len(c.WorkloadIdentityClientID) > 0 {
+		return getAuthorizerForWorkloadIdentity(ctx, c.WorkloadIdentityClientID, c.WorkloadIdentityToken, resource, aadEndpoint, tenantID)
+	}
+
+	return nil, fmt.Errorf("no valid identity access mode specified")
+}
+
+func getAuthorizerForWorkloadIdentity(ctx context.Context, clientID, signedAssertion, resource, aadEndpoint, tenantID string) (autorest.Authorizer, error) {
+	cred, err := confidential.NewCredFromAssertion(signedAssertion)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create confidential creds: %w", err)
+	}
+	confidentialClientApp, err := confidential.New(clientID, cred,
+		confidential.WithAuthority(fmt.Sprintf("%s%s/oauth2/token", aadEndpoint, tenantID)))
+	if err != nil {
+		return nil, fmt.Errorf("failed to create confidential client app: %w", err)
+	}
+	scope := strings.TrimSuffix(resource, "/")
+	// .default needs to be added to the scope
+	if !strings.HasSuffix(resource, ".default") {
+		scope += "/.default"
+	}
```

Comment:
if !strings.Contains(resource, ".default") {
   scope := fmt.Sprintf("%s/.default", resource)
}

Reply:
The resource could have `/` suffix or not, so we're sanitizing it `scope := strings.TrimSuffix(resource, "/")` before making a deterministic append. `HasSuffix` checks exactly what we need, wouldn't `strings.Contains` fail for any url that has `.default` as part of the resource url?

Reply:
`.default` generally appears only in scope which is pretty standard. I have not seen it in any resource strings containing same. My goal behind this approach is to avoid multiple strings manipulation.